src/adaptors/amqp/amqp_adaptor.c

```diff
@@ -1257,29 +1238,59 @@ static int AMQP_link_detach_handler(qd_router_t *router, qd_link_t *link, qd_det
         }
     }
 
-    qdr_link_t     *rlink  = (qdr_link_t*) qd_link_get_context(link);
-    pn_condition_t *cond   = qd_link_pn(link) ? pn_link_remote_condition(qd_link_pn(link)) : 0;
+    // Notify the core that a detach has been received.
 
+    qdr_link_t *rlink = (qdr_link_t *) qd_link_get_context(link);
     if (rlink) {
-        //
-        // If this is the second (response) detach or the link hasn't really detached but is being dropped due to parent
-        // connection/session loss then this is the last proton event that will be generated for this link. The qd_link
-        // will be freed on return from this call so remove the cross linkage between it and the qdr_link peer.
-
-        if (dt == QD_LOST || qdr_link_get_context(rlink) == 0) {
-            // note qdr_link context will be zeroed when the core sends the first detach, so if it is zero then this is
-            // the second detach!
-            qd_link_set_context(link, 0);
-            qdr_link_set_context(rlink, 0);
-        }
-
-        qdr_error_t *error = qdr_error_from_pn(cond);
-        qdr_link_detach(rlink, dt, error);
+        pn_condition_t *cond  = pn_link_remote_condition(pn_link);
+        qdr_error_t    *error = qdr_error_from_pn(cond);
+        qdr_link_detach_received(rlink, error);
+    } else if ((pn_link_state(pn_link) & PN_LOCAL_CLOSED) == 0) {
+        // Normally the core would be responsible for sending the response detach to close the link (via
+        // CORE_link_detach) but since there is no core link that will not happen.
+        pn_link_close(pn_link);
     }
 
     return 0;
 }
 
+
+/**
+ * Link closed handler
+ *
+ * This is the last callback for the given link - the link will be freed on return from this call! Forced is true if the
+ * link has not properly closed (detach handshake completed).
+*/
+static void AMQP_link_closed_handler(qd_router_t *router, qd_link_t *qd_link, bool forced)
```

Comment:
I was wondering if the `AMQP_` prefix should be used only for functions that are called directly as a response to proton AMQP events. This function can be called `qd_link_close()` ?

Reply:
Could be... but I'm not sure if other devs would be ok with that new naming concept.  @ssorj @ted-ross and others - opinions?